Terminal User Interface (TUI) for kubefwd

cjimti opened the pull request:
Adds an interactive terminal UI mode to kubefwd, providing real-time visualization of port forwards, bandwidth metrics, and pod logs. Enable with `--tui` flag.

```bash
sudo -E kubefwd svc -n my-namespace --tui
```

<img width="984" height="856" alt="SCR-20251224-bjnw" src="https://github.com/user-attachments/assets/0119f842-cd41-4204-8dab-25395fcf950b" />

<img width="987" height="857" alt="SCR-20251224-bkaj" src="https://github.com/user-attachments/assets/4528435b-3e51-4f6a-9946-3ad841500f8e" />

<img width="988" height="846" alt="SCR-20251224-bkfx" src="https://github.com/user-attachments/assets/28b3b197-1e93-47ab-9c75-aa7945a04dd9" />


## New Features

### TUI Mode (`--tui`)

A full-screen terminal interface built with [Bubble Tea](https://github.com/charmbracelet/bubbletea) that displays:

- **Services table**: All active port forwards with hostname, port, status, and bandwidth metrics
- **Logs panel**: Application logs with auto-follow/pause functionality
- **Detail view**: Per-service breakdown with connection info, bandwidth sparklines, HTTP request logs, and live pod log streaming
- **Status bar**: Summary counts and aggregate bandwidth

### Auto-Reconnect (`--auto-reconnect`, `-a`)

Automatically reconnects port forwards when connections are lost. Uses exponential backoff (1s to 5min) to avoid hammering the API during cluster issues.

### Real-time Metrics

- Bytes sent/received per port forward
- Transfer rates (B/s) with rolling averages
- HTTP request sniffing (method, path, status, duration, size)
- Historical rate data for sparkline visualization

## Keyboard Shortcuts

### Main View
| Key | Action |
|-----|--------|
| `j/k` or arrows | Navigate services |
| `Enter` | Open detail view |
| `Tab` | Switch focus (Services/Logs) |
| `/` | Filter services |
| `b` | Toggle bandwidth columns |
| `c` | Toggle compact/full view |
| `r` | Reconnect errored services |
| `?` | Help |
| `q` | Quit |

### Detail View
| Key | Action |
|-----|--------|
| `Tab` | Switch tabs (Info/HTTP/Logs) |
| `1-9` | Copy connect string to clipboard |
| `j/k` | Scroll |
| `Esc` | Close |

### Logs Panel
| Key | Action |
|-----|--------|
| `j/k` | Scroll (pauses auto-follow) |
| `G` | Go to bottom (resumes follow) |

## Architecture

### New Packages

**`pkg/fwdtui/`** - Main TUI package
- `ui.go` - Root model, event routing, layout management
- `commands.go` - Bubble Tea commands for async operations
- `messages.go` - Message types for inter-component communication

**`pkg/fwdtui/components/`** - UI components
- `services.go` - Services table (uses [bubble-table](https://github.com/evertras/bubble-table))
- `detail.go` - Service detail view with tabs
- `logs.go` - Scrollable log viewport with follow/pause
- `header.go` - Version header
- `statusbar.go` - Footer with stats
- `help.go` - Help overlay
- `sparkline.go` - ASCII sparkline renderer

**`pkg/fwdtui/state/`** - State management
- `store.go` - Central state store for forwards and metrics
- `snapshot.go` - Immutable snapshot types
- `history.go` - Rate history for sparklines

**`pkg/fwdtui/events/`** - Event system
- `bus.go` - Pub/sub event bus
- `types.go` - Event type definitions

**`pkg/fwdtui/styles/`** - UI styling
- `styles.go` - Lipgloss style definitions

**`pkg/fwdmetrics/`** - Metrics collection
- `registry.go` - Global metrics registry with periodic sampling
- `stream.go` - Per-connection byte counters
- `connection.go` - Wraps net.Conn for metrics
- `dialer.go` - Wraps dialer for automatic instrumentation
- `rate.go` - Rate calculation with rolling windows
- `httpsniff.go` - HTTP request/response detection
- `types.go` - Snapshot types for metrics export

### Modified Files

- `cmd/kubefwd/services/services.go` - Added `--tui` and `--auto-reconnect` flags, TUI lifecycle management, pod log streaming support
- `pkg/fwdport/fwdport.go` - Integrated metrics tracking, emit TUI events on status changes
- `pkg/fwdservice/fwdservice.go` - Added auto-reconnect logic, container name tracking for log streaming
- `pkg/fwdsvcregistry/fwdsvcregistry.go` - Emit events on service add/remove, added reconnect API

## New Dependencies

```
github.com/charmbracelet/bubbletea v1.2.4
github.com/charmbracelet/bubbles v0.20.0
github.com/charmbracelet/lipgloss v1.0.0
github.com/evertras/bubble-table v0.18.0
github.com/muesli/termenv v0.15.2
```

## Testing

- `pkg/fwdtui/tui_test.go` - Basic store and event bus tests
- `pkg/fwdmetrics/metrics_flow_test.go` - Metrics collection and rate calculation tests

## Screenshots

The TUI provides a split-pane view:
- Top: Services table showing all active port forwards
- Bottom: Application logs with following/paused indicator
- Press Enter on any service to see detailed bandwidth graphs, HTTP logs, and stream pod logs

## Migration Notes

No breaking changes. The `--tui` flag is opt-in. Default behavior (log-based output) remains unchanged.

## Known Limitations

- UDP port forwarding not supported (Kubernetes API limitation)
- HTTP sniffing is best-effort and may miss some requests
- Pod log streaming requires the pod to still be running

TWRaven commented on README.md:
```diff
@@ -25,7 +25,7 @@ When working on our local workstation, my team and I often build applications th
 <div align="center">
   <img width="654" height="684" src="https://mk.imti.co/images/content/kubefwd-net.png" alt="kubefwd - Kubernetes Port Forward Diagram">
-</div>
+</div>no
```
unintended change?

TWRaven commented on pkg/fwdmetrics/rate.go:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	rateIn = float64(curr.BytesIn-prev.BytesIn) / duration
+	rateOut = float64(curr.BytesOut-prev.BytesOut) / duration
```
```suggestion
    // Handle counter wraparound or reset
    if curr.BytesIn < prev.BytesIn || curr.BytesOut < prev.BytesOut {
        return 0, 0
    }

	rateIn = float64(curr.BytesIn-prev.BytesIn) / duration
	rateOut = float64(curr.BytesOut-prev.BytesOut) / duration
```

If a pod is restarted, could it be possible that the new bytes counter is smaller than the old one? If so, this safety check will help.

TWRaven commented on pkg/fwdmetrics/rate.go:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	rateIn = float64(newest.BytesIn-oldest.BytesIn) / duration
+	rateOut = float64(newest.BytesOut-oldest.BytesOut) / duration
```
```suggestion
    // Handle counter wraparound or reset
    if newest.BytesIn < oldest.BytesIn || newest.BytesOut < oldest.BytesOut {
        return 0, 0
    }

	rateIn = float64(curr.BytesIn-prev.BytesIn) / duration
	rateOut = float64(curr.BytesOut-prev.BytesOut) / duration
```

If a pod is restarted, could it be possible that the new bytes counter is smaller than the old one? If so, this safety check will help.

cjimti commented:
Thanks @TWRaven for the review.
  - Fixed typo in README
  - Added comment for scheduleReconnect
  - Added counter wraparound protection in rate.go
  - Fixed race conditions in registry.go  
  - Made BytesIn/BytesOut unexported in types.go

cjimti commented on pkg/fwdmetrics/registry.go:
```diff
@@ -0,0 +1,178 @@
+		// Create service if it doesn't exist
+		svc = NewServiceMetrics(pf.ServiceName, pf.Namespace, pf.Context)
+		r.RegisterService(svc)
+	}
```
fixed. 

now using a write lock for the entire check-and-create to eliminate the race. I applied same pattern to UnregisterPortForward too.

cjimti commented on pkg/fwdtui/events/bus.go:
```diff
@@ -0,0 +1,108 @@
+	// Call type-specific handlers
+	if handlers, ok := b.handlers[event.Type]; ok {
+		for _, handler := range handlers {
+			handler(event)
```
good point! I'll add a recover wrapper.

cjimti commented:
Stabilized reconnection flow: https://vimeo.com/1149248063
